include/boost/math/tools/polynomial.hpp

```diff
@@ -608,6 +616,117 @@ polynomial<T> operator - (polynomial<T> a)
     return a;
 }
 
+
+/* From Knuth, 4.6.1:
+ * 
+ * We may write any nonzero polynomial u(x) as
+ *
+ *      u(x) = cont(u) · pp(u(x))
+ *
+ * where cont(u), the content of u, is an element of S, and pp(u(x)), the primitive
+ * part of u(x), is a primitive polynomial over S. 
+ * When u(x) = 0, it is convenient to define cont(u) = pp(u(x)) = O.
+ */
+
+template <class T>
+T content(polynomial<T> const &x)
+{
+    polynomial<T> const zero = zero_element(std::multiplies< polynomial<T> >());
+    return x == zero ? T(0) : gcd_n(x.data().begin(), x.data().end());
+}
+
+// Knuth, 4.6.1
+template <class T>
+polynomial<T> primitive_part(polynomial<T> const &x)
+{
+    polynomial<T> const zero = zero_element(std::multiplies< polynomial<T> >());
+    return x == zero ? zero : x / content(x);
+}
+
+
+// Trivial but useful convenience function defined by Knuth.
+template <class T>
+T leading_coefficient(polynomial<T> const &x)
+{
+    return x.data().back();
+}
+
+namespace detail
+{
+    template <class T>
+    T reduce_to_primitive(polynomial<T> &u, polynomial<T> &v)
+    {
+        T const u_cont = content(u), v_cont = content(v);
+        u /= u_cont;
+        v /= v_cont;
+        return gcd(u_cont, v_cont);
+    }
+}
+
+/* 4.6.1C: Greatest common divisor over a unique factorization domain.
+ * 
+ * Although step C3 keeps the coefficients to a "reasonable" size, they are
+ * still potentially several binary orders of magnitude larger than the inputs.
+ */
+template <class T>
+typename enable_if_c< std::numeric_limits<T>::is_integer, polynomial<T> >::type
+gcd_ufd(polynomial<T> const &a, polynomial<T> const &b)
+{
+    BOOST_ASSERT(a);
+    BOOST_ASSERT(b);
+    typedef typename polynomial<T>::size_type N;
+    // If T is a POD, double the output size, otherwise assume it is multi-prec.
+    // This precludes the use of long as a coefficient type in combination with
+    // gcd.
+    /*
+    typedef typename mpl::if_<is_pod<T>,
+                            typename mpl::if_<is_unsigned<T>, 
+                                    typename uint_t<sizeof(T) * 16>::fast, 
+                                    typename int_t<sizeof(T) * 16>::fast >::type,
+                            T>::type U;
+                            */
+    typedef T U;
+    using boost::math::gcd;
+    
+    polynomial<U> u(a), v(b);
+    U const d = detail::reduce_to_primitive(u, v);
+    U g = 1, h = 1;
+    polynomial<U> r;
+    while (true)
+    {
+        // Pseudo-division.
+        N const delta = u.degree() - v.degree();
+        r = u % v;
+        if (!r)
+            goto gcd_end;
+        if (r.degree() == 0)
+        {
+            v = polynomial<U>(U(1));
+            goto gcd_end;
```

Comment:
Yes although I left the goto in because I wasn't immediately sure how to
take it out without double-checking the termination condition. And I find
anything taboo strangely fascinating. :)
On 21 Apr 2016 3:15 AM, "jzmaddock" <notifications@github.com> wrote:

> In include/boost/math/tools/polynomial.hpp
> <https://github.com/boostorg/math/pull/34#discussion_r60449716>:
>
> > +
> > +    polynomial<U> u(a), v(b);
> > +    U const d = detail::reduce_to_primitive(u, v);
> > +    U g = 1, h = 1;
> > +    polynomial<U> r;
> > +    while (true)
> > +    {
> > +        // Pseudo-division.
> > +        N const delta = u.degree() - v.degree();
> > +        r = u % v;
> > +        if (!r)
> > +            goto gcd_end;
> > +        if (r.degree() == 0)
> > +        {
> > +            v = polynomial<U>(U(1));
> > +            goto gcd_end;
>
> Is it possible to write this without a goto?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/boostorg/math/pull/34/files/d484fa657c65cc8a233e1aeabb72c2e7199c581e#r60449716>
>